src/Microsoft.ML.Data/DataView/InternalSchemaDefinition.cs

```diff
@@ -159,50 +161,33 @@ public static void GetVectorAndKind(MemberInfo memberInfo, out bool isVector, ou
 
-        /// <summary>
-        /// Given a parameter info on a type, returns whether this appears to be a vector type,
-        /// and also the associated data kind for this type. If a data kind could not
-        /// be determined, this will throw.
-        /// </summary>
-        /// <param name="parameterInfo">The parameter info to inspect.</param>
-        /// <param name="isVector">Whether this appears to be a vector type.</param>
-        /// <param name="kind">The data kind of the type, or items of this type if vector.</param>
-        public static void GetVectorAndKind(ParameterInfo parameterInfo, out bool isVector, out DataKind kind)
-        {
-            Contracts.AssertValue(parameterInfo);
-            Type rawParameterType = parameterInfo.ParameterType;
-            var name = parameterInfo.Name;
-            GetVectorAndKind(rawParameterType, name, out isVector, out kind);
-        }
-
         /// <summary>
         /// Given a type and name for a variable, returns whether this appears to be a vector type,
-        /// and also the associated data kind for this type. If a data kind could not
+        /// and also the associated data type for this type. If a valid data type could not
         /// be determined, this will throw.
         /// </summary>
         /// <param name="rawType">The type of the variable to inspect.</param>
         /// <param name="name">The name of the variable to inspect.</param>
         /// <param name="isVector">Whether this appears to be a vector type.</param>
-        /// <param name="kind">The data kind of the type, or items of this type if vector.</param>
-        public static void GetVectorAndKind(Type rawType, string name, out bool isVector, out DataKind kind)
+        /// <param name="itemType">
+        /// For non-vectors, this is set to <paramref name="rawType"/>.
+        /// For vectors, this is set to the type of the items stored as values in the vector.
+        /// </param>
+        public static void GetVectorAndItemType(Type rawType, string name, out bool isVector, out Type itemType)
         {
             // Determine whether this is a vector, and also determine the raw item type.
-            Type rawItemType;
             isVector = true;
             if (rawType.IsArray)
-                rawItemType = rawType.GetElementType();
+                itemType = rawType.GetElementType();
             else if (rawType.IsGenericType && rawType.GetGenericTypeDefinition() == typeof(VBuffer<>))
-                rawItemType = rawType.GetGenericArguments()[0];
+                itemType = rawType.GetGenericArguments()[0];
             else
             {
-                rawItemType = rawType;
+                itemType = rawType;
                 isVector = false;
             }
 
-            // Get the data kind, and the item's column type.
-            if (rawItemType == typeof(string))
-                kind = DataKind.Text;
-            else if (!rawItemType.TryGetDataKind(out kind))
```

Comment:
This was actually important -- we do not *want* `string` associated with the type `TextType`, because `TextType.Instance.RawType` is not `string`, it is `ReadOnlyMemory<char>`. We need to handle that type mapping "specially" just as we specially map arrays to `VBuffer` in other code here, and so on, and so on. It is not the job of `ColumnType` to be aware of the details of how in this convenience API we perform mappings from user types into `IDataView` land. So if that is why I see those tests against `string` that I commented on earlier as being odd, let's not do that.

Reply:
Good call. I've added this back, by doing the `string` => `ROM<char>` type conversion here.

---
In reply to: [248940771](https://github.com/dotnet/machinelearning/pull/2176#discussion_r248940771) [](ancestors = 248940771)